lib/dfdutils/dfd2vk.inl

```diff
@@ -71,133 +80,170 @@ if (KHR_DFDVAL(dfd + 1, MODEL) == KHR_DF_MODEL_RGBSDA) {
         if ((r & i_NORMALIZED_FORMAT_BIT) && (r & i_SIGNED_FORMAT_BIT)) return VK_FORMAT_A2B10G10R10_SNORM_PACK32;
         if (!(r & i_NORMALIZED_FORMAT_BIT) && !(r & i_SIGNED_FORMAT_BIT)) return VK_FORMAT_A2B10G10R10_UINT_PACK32;
         if (!(r & i_NORMALIZED_FORMAT_BIT) && (r & i_SIGNED_FORMAT_BIT)) return VK_FORMAT_A2B10G10R10_SINT_PACK32;
-      } else if (R.size == 11) return VK_FORMAT_B10G11R11_UFLOAT_PACK32;
+      } else if (R.size == 11) {
+          return VK_FORMAT_B10G11R11_UFLOAT_PACK32;
+      } else if (R.size == 10 && G.size == 10 && B.size == 0) { 
+          return VK_FORMAT_R10X6G10X6_UNORM_2PACK16;
+      } else if (R.size == 12 && G.size == 12 && B.size == 0) { 
+          return VK_FORMAT_R12X4G12X4_UNORM_2PACK16;
+      }
+    } else if (wordBytes == 8) { /* 4PACK16 */
+      if (r & i_YUVSDA_FORMAT_BIT) {
+        if (R.size == 10 && R.offset == 6 && A.size == 10 && A.offset == 38)
+          return  VK_FORMAT_G10X6B10X6G10X6R10X6_422_UNORM_4PACK16;
```

Comment:
I'm very rusty, but I don't think this is right. 4PACK16 means four values each of which should be treated as 16-bit packed; the 16-bit units are in left-to-right order with increasing address as with other Vulkan and GL formats which list the channels independently, but then the contents of those 16-bit units are listed left-to-right as with packed formats (because of how big-endian numbers are written if you're drawing out the bit assignments).

VK_FORMAT_G10X6B10X6G10X6R10X6_422_UNORM_4PACK16 is 4 16-bit packed values:

G10X6 in the 16 bits starting at address offset 0 (in bytes).
B10X6 in the 16 bits starting at address offset 2 (in bytes).
G10X6 in the 16 bits starting at address offset 4 (in bytes).
R10X6 in the 16 bits starting at address offset 6 (in bytes).

Within those 16-bit quantities (which can have the bytes ordered by the implementation's endianness), the X channel takes up the low 6 bits, and the colour channel takes up the top ten bits.

In other words, for a little-endian implementation and the KDFS convention of counting all the offsets starting at bit 0 of byte 0, I believe VK_FORMAT_G10X6B10X6G10X6R10X6_422_UNORM_4PACK16 matches:

R.size == 10 && R.offset == 54 && A.size == 10 && A.offset == 38.

(TL;DR: swap the conditions with the format below.)

Incidentally, might be nice to add a reminder comment that the second G channel is getting packed into alpha for the purposes of this translation. :-)

Reply:
I agree with your analysis of these packed types. Not sure how I got those R offsets so wrong. There must be errors in the vk2dfd as well since the round trip tests are all working. Did you review vk2dfd and not spot anything or did you not review it?

Reply:
Looks like you read them all as a single packed (32- or 64-bit) quantity. I did look at vk2dfd, and I just had another look, and don't think it looks wrong. I'll keep looking in case I'm missing something.

Reply:
Whoops!!! The confusion here is the different mapping of YUV to RGB used by Vulkan and KDFS. 

`G10X6B10X6G10X6R10X6` for Vulkan is YUYV which in terms of KDFS  channel id values in `khr_df.h` is RGRB and in the output from `interpretDFD` is RGAB. Hence my original offsets were correct. I shall revert my local changes and add a more detailed comment about the mapping.

@fluppeteer please check if you agree.

Reply:
*Oh*. Sorry, because KHR_DF_CHANNEL_RGBSDA_R = KHR_DF_CHANNEL_YUVSDA_Y (etc.) - and we're assuming that it's being interpreted as YUYV, not as the underlying GBGR (which it doesn't have to be).

That makes sense; I've not re-checked all your numbers, but it seems plausible. I'd advocate using aliased enum values so you can actually refer to the channels in YCbCr/YUV terms for these formats, and noting that we're assuming the YCBCR conversion (although if there's no samplerYCbCrconversion object then that still makes packed 444 ambiguous, because it looks like RGB).

Sorry for causing confusion.

Reply:
> I'd advocate using aliased enum values

Unfortunately it is not an enumeration here, it is the parameters passed to `interpretDFD`. I'll experiment with how to make it more clear. #defines and unions come to mind.

Reply:
I used #defines. Much clearer.